### PR TITLE
feat(frontend): Use secondary index on logical scan with required order

### PR DESCRIPTION
**Pull request by GeorgKreuzmayr**
Signed-off-by: Georg Kreuzmayr <g.kreuzmayr@gmail.com>

This Pull request introduces the usage of a lookup join when the required order can be satisfied using an index scan, but the index columns do not cover all required output columns.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/batch_project.rs [hunk 1]
```diff
@@ -48,6 +49,15 @@ impl BatchProject {
         BatchProject { base, logical }
     }
 
+    pub fn new_with_order(logical: LogicalProject, order: Order) -> Self {
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/batch_project.rs:52]: can we remove this? I think for the BatchLookUp join is ok to give it a order if we Know the order is that **in fact** and the order derivation with its input is a little hard. but for the project its order should be determined by its input :thinking:

File: src/frontend/src/optimizer/plan_node/logical_scan.rs [hunk 2]
```diff
@@ -523,7 +542,11 @@ impl LogicalScan {
                     })
                     .collect(),
             }
-            .satisfies(required_order)
+            .satisfies(
+                &self
+                    .o2i_col_mapping()
+                    .rewrite_provided_order(required_order),
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/logical_scan.rs:548]:
```suggestion
                    .rewrite_required_order(required_order),
```
you can read the comments on these two functions and see their slight difference.

File: src/frontend/src/optimizer/plan_node/batch_project.rs (hunk 1 above)
Comment (GeorgKreuzmayr) [reply src/frontend/src/optimizer/plan_node/batch_project.rs]: Yeah, sorry this was my first approach, where I added a projection after the lookup join and add the order injection there. This is of course irrelevant now.

File: e2e_test/batch/order/test_order_requirers_index.part [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+statement ok
+drop index idx1;
+
+# Index includes all required columns
+
+statement ok
+create index idx1 on t1(b) include(a, c);
+
+query I rowsort
+select * from t1 order by b;
```
Comment (GeorgKreuzmayr) [inline e2e_test/batch/order/test_order_requirers_index.part:80]: There seems to be a problem in this case, as the output is not sorted by b.
Is it possible that this is an error in the implementation, or did I make wrong assumptions here?
@st1page

File: src/frontend/planner_test/tests/testdata/index_selection.yaml [hunk 4]
```diff
@@ -60,6 +52,51 @@
...
+    BatchExchange { order: [idx1.a ASC], dist: Single }
+      BatchScan { table: idx1, columns: [idx1.a, idx1.b, idx1.c], distribution: UpstreamHashShard(idx1.a, idx1.b, idx1.c) }
+- sql: |
+    /* Index includes all required columns I */
+    create table t1 (a int, b int, c int);
+    create index idx1 on t1(b) include(c, a);
+    select * from t1 order by b
+  batch_plan: |
+    BatchExchange { order: [idx1.b ASC], dist: Single }
+      BatchScan { table: idx1, columns: [idx1.a, idx1.b, idx1.c], distribution: UpstreamHashShard(idx1.b) }
```
Comment (GeorgKreuzmayr) [inline src/frontend/planner_test/tests/testdata/index_selection.yaml:70]: The same setup that causes problems in the e2e test is also shown here.
Somehow the index scan does not provide the order in this case. @st1page

**Pull request closed without merging**